### PR TITLE
internals: Implement Serde traits for ArrayVec

### PR DESCRIPTION
**Pull request by Abeeujah**
This implementation is a prerequisite for the serde traits impl for `WitnessVersion` and `WitnessProgram`.

Add tests covering round-trip serialization, capacity overflow errors, and wire-format compatibilitywith slices using both JSON and Bincode.

references #3513

### CODE REVIEW

File: internals/src/array_vec.rs [hunk 1]
```diff
@@ -178,6 +178,75 @@ impl<T: Copy + core::hash::Hash, const CAP: usize> core::hash::Hash for ArrayVec
...
+
+                if let Some(hint) = seq.size_hint() {
+                    if hint > CAP {
+                        return Err(Error::invalid_length(hint, &self));
+                    }
+                }
+
+                let mut out = ArrayVec::<T, CAP>::new();
+                while let Some(elem) = seq.next_element::<T>()? {
+                    if out.len() >= CAP {
```
Comment (tcharding) [inline internals/src/array_vec.rs:239]:
```suggestion
                    // The `push()` call below panics if array is full but we want to error.
                    if out.len() >= CAP {
```

File: internals/src/array_vec.rs [hunk 2]
```diff
@@ -210,6 +279,101 @@ mod tests {
...
+    #[cfg(feature = "test-serde")]
+    #[test]
+    fn serde_round_trip_u8() {
+        let mut av = ArrayVec::<u8, 8>::new();
+        av.extend_from_slice(b"abc");
+
+        let json = crate::serde_json::to_string(&av).expect("serde_json failed to encode");
+        let decoded: ArrayVec<u8, 8> =
+            crate::serde_json::from_str(&json).expect("serde_json failed to decode");
+        assert_eq!(av, decoded);
```
Comment (tcharding) [inline internals/src/array_vec.rs]: Can we have these in the order 'got, want' - this is by convention so that it is easier for future debuggers.

For a mnemonic, this is in alphabetic order, as is `actual, expected`.

**Comment by tcharding** [comment]
Nice PR. Reviewed: 680cf025e68d0bb4cffeb86987e895057c3a6ed8

File: internals/src/array_vec.rs [hunk 3]
```diff
@@ -210,6 +280,101 @@ mod tests {
...
+
+        let json = crate::serde_json::to_string(&want).expect("serde_json failed to encode");
+        let got: ArrayVec<u8, 8> =
+            crate::serde_json::from_str(&json).expect("serde_json failed to decode");
+        assert_eq!(want, got);
+
+        let bin = crate::bincode::serialize(&want).expect("bincode failed to encode");
+        let got: ArrayVec<u8, 8> =
+            crate::bincode::deserialize(&bin).expect("bincode failed to decode");
+        assert_eq!(want, got);
```
Comment (tcharding) [inline internals/src/array_vec.rs]: This is not alphabetic order.

**Review by tcharding — APPROVED** [review]
ACK 62895cdbf2fba00d3306682102b9ac269edf003c

**Comment by Abeeujah** [comment]
ping @apoelstra

**Comment by apoelstra** [comment]
Oops, sorry, this failed because last week I had the `generate-files.sh` check turned on (which was failing on master, and therefore on this PR) and I didn't notice it due to travel.

utACK. Let me restart the tests.

**Review by apoelstra — APPROVED** [review]
ACK 62895cdbf2fba00d3306682102b9ac269edf003c; successfully ran local tests

**Comment by Abeeujah** [comment]
Sorry, It's been a week now, and the merge script hasn't picked this up, could it be something on my end I should look into?

**Comment by apoelstra** [comment]
Sorry @Abeeujah, no, I've just been travelling and this fell through the cracks.

**Comment by Abeeujah** [comment]
Ooh word, Sorry @apoelstra this got to bother you during your travels, do enjoy your time, wishing you safe travels!

File: internals/src/array_vec.rs [hunk 4]
```diff
@@ -178,6 +178,76 @@ impl<T: Copy + core::hash::Hash, const CAP: usize> core::hash::Hash for ArrayVec
...
+        S: crate::serde::Serializer,
+    {
+        use crate::serde::ser::SerializeSeq;
+
+        let mut seq = serializer.serialize_seq(Some(self.len()))?;
+        for elem in self.as_slice() {
+            seq.serialize_element(elem)?;
+        }
+        seq.end()
+    }
```
Comment (Kixunil) [inline internals/src/array_vec.rs:196]: Shouldn't this detect that `T == u8` and call into `serialize_bytes`?

File: internals/src/array_vec.rs [hunk 5]
```diff
@@ -178,6 +178,76 @@ impl<T: Copy + core::hash::Hash, const CAP: usize> core::hash::Hash for ArrayVec
...
+
+                if let Some(hint) = seq.size_hint() {
+                    if hint > CAP {
+                        return Err(Error::invalid_length(hint, &self));
+                    }
+                }
+
+                let mut out = ArrayVec::<T, CAP>::new();
+                while let Some(elem) = seq.next_element::<T>()? {
+                    // The `push()` call below panics if array is full but we want to error.
```
Comment (Kixunil) [inline internals/src/array_vec.rs:238]: Might be time to add `try_push` and change `push` to call `try_push(element).unwrap()`

**Review by Kixunil — COMMENTED** [review]
I don't think this is really prerequisite but it does make things nicer, so concept ACK.

**Review by suthephbuytengchay-ctrl — APPROVED** [review]

**Comment by tcharding** [comment]
If you farm activity in this repo you'll be blocked @suthephbuytengchay-ctrl

File: internals/src/array_vec.rs (hunk 4 above)
Comment (Abeeujah) [reply internals/src/array_vec.rs:196]: I'm thinking, as per all other `serialize_bytes` usage in the workspace, this would be easier if there's a method that converts the `ArrayVec` to bytes.

File: internals/src/array_vec.rs (hunk 5 above)
Comment (Abeeujah) [reply internals/src/array_vec.rs:238]: Yeah, would be a good addition, but in this instance, doesn't the `unwrap` still hit the `panic!` we're trying to return an `Err` for?

File: internals/src/array_vec.rs (hunk 4 above)
Comment (apoelstra) [reply internals/src/array_vec.rs:196]: We should match what `Vec<u8>` does. Lemme take a look..

Comment (apoelstra) [reply internals/src/array_vec.rs:196]: @Kixunil `Vec<T>` just calls `serializer.collect_seq()`, which is a method I just learned about now and is equivalent to the above code.

Unfortunately `serialize_seq`/`serialize_element` does different things than `serialize_bytes`, so if we were to take your suggestion, even though it'd be unambiguously better, we would then be doing something different from the serde impls of all the stdlib collection types. So then it'd be a serialization-breaking change for users to switch from `Vec<T>` to `ArrayVec<T>`.

As an optimization/code cleanliness thing we should probably grep the codebase for every instance of `serialize_element`, which probably always occurs in this pattern, and replace the manual loop with `collect_seq`.

File: internals/src/array_vec.rs (hunk 5 above)
Comment (apoelstra) [reply internals/src/array_vec.rs:238]: Yeah, I'd prefer that. But ofc can be done in a followup.

Comment (Kixunil) [reply internals/src/array_vec.rs:238]: @Abeeujah I meant to call `try_push` here instead and convert the error as invalid length.

File: internals/src/array_vec.rs (hunk 4 above)
Comment (Kixunil) [reply internals/src/array_vec.rs:196]: Good point, since this is effectively supposed to be a replacement of `Vec`.

But what I think we should be careful about is wherever we define serialization and we know there are actually bytes we should prefer bytes.

Notably, I'm not confident that `WitnessProgram` has any well-defined serialization (note that our `WitnessProgram` is not really witness program) and we probably shouldn't offer one. However if we had actual `WitnessProgram` then I think bytes/hex conditioned on the format being human-readable makes most sense.
